PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,491 @@
+    public func set(latestNetworkAndAdvertisingIdsSent: [String: Any], appUserID: String) {
+        writeCache {
+            self.userDefaults.setValue(latestNetworkAndAdvertisingIdsSent,
+                                       forKey: CacheKeyBases.attributionDataDefaults + appUserID)
+        }
+    }
+
+    @objc open func clearLatestNetworkAndAdvertisingIdsSent(appUserID: String) {
+        writeCache {
+            self.userDefaults.removeObject(forKey: CacheKeyBases.attributionDataDefaults + appUserID)
+        }
+    }
+
+    private func cacheDurationInSeconds(isAppBackgrounded: Bool) -> Double {
+        return isAppBackgrounded ? cacheDurationInSecondsInBackground : cacheDurationInSecondsInForeground
+    }
+
+    // MARK: - Helper functions
+
+    // Uses the accessQueue to synchronously write to the UserDefaults.
+    // Warning: do NOT nest calls to `writeCache`, it will deadlock.
+    // Only make calls to functions starting with "threadUnsafe" as those are guaranteed to not have any locking
+    // mechanisms.
+    private func writeCache(block: @escaping () -> Void) {
+        // .barrier is not needed here because we're using `.sync` instead of the normal .async multi-reader
+        // single-writer dispatch queue synchronization pattern. However, because we're using this as a way
+        // to control access to the datasource, passing `.barrier` flag is meant to help signal that intent.
+        accessQueue.sync(flags: .barrier) {
+            block()
+
+            // While Apple states `this method is unnecessary and shouldn't be used`
+            // https://developer.apple.com/documentation/foundation/userdefaults/1414005-synchronize
+            // It didn't become unnecessary until iOS 12 and macOS 10.14 (Mojave):
+            // https://developer.apple.com/documentation/macos-release-notes/foundation-release-notes
+            // there are reports it is still needed if you save to defaults then immediately kill the app.
+            // Also, it has not been marked deprecated... yet.
+            self.userDefaults.synchronize()
+        }
+    }
+
+    // Uses the accessQueue to synchronously read from the UserDefaults which ensures access is threadsafe.
+    private func readCache<T>(block: @escaping () -> T) -> T {
+        return accessQueue.sync(execute: block)
+    }
+
+    // MARK: - Testing
+
+    public static func newAttribute(dictionary: [String: Any]) -> SubscriberAttribute {
+        // swiftlint:disable force_cast
+        let key = dictionary[SubscriberAttribute.keyKey] as! String
+        let value = dictionary[SubscriberAttribute.valueKey] as? String
+        let isSynced = (dictionary[SubscriberAttribute.isSyncedKey] as! NSNumber).boolValue
+        let setTime = dictionary[SubscriberAttribute.setTimeKey] as! Date
+        // swiftlint:enable force_cast
+
+        return SubscriberAttribute(withKey: key, value: value, isSynced: isSynced, setTime: setTime)
+    }
```

Comment:
why is this needed? 
if we need something for testing we should add it to the test targets instead

Reply:
Obsolete comment, it's used internally 🙃  removed comment and changed access level to private.